crates/red_knot_python_semantic/src/types/infer.rs

```diff
@@ -840,4 +983,263 @@ b = z
+                class A: pass
+                import b
+                class C(b.B): pass
+            ",
+        )?;
+        db.write_dedented(
+            "src/b.py",
+            "
+                from a import A
+                class B(A): pass
+            ",
+        )?;
+
+        let a = system_path_to_file(&db, "src/a.py").expect("Expected file to exist.");
+        let c_ty = module_global_symbol_ty_by_name(&db, a, "C");
+        let Type::Class(c_class) = c_ty else {
+            panic!("C is not a Class")
+        };
+        let c_bases = c_class.bases(&db);
+        let b_ty = c_bases.first().unwrap();
+        let Type::Class(b_class) = b_ty else {
+            panic!("B is not a Class")
+        };
+        assert_eq!(b_class.name(&db), "B");
+        let b_bases = b_class.bases(&db);
+        let a_ty = b_bases.first().unwrap();
+        let Type::Class(a_class) = a_ty else {
+            panic!("A is not a Class")
+        };
+        assert_eq!(a_class.name(&db), "A");
+
+        Ok(())
+    }
+
+    #[test]
+    fn local_inference() -> anyhow::Result<()> {
+        let mut db = setup_db();
+
+        db.write_file("/src/a.py", "x = 10")?;
+        let a = system_path_to_file(&db, "/src/a.py").unwrap();
+
+        let parsed = parsed_module(&db, a);
+
+        let statement = parsed.suite().first().unwrap().as_assign_stmt().unwrap();
+        let model = SemanticModel::new(&db, a);
+
+        let literal_ty = statement.value.ty(&model);
+
+        assert_eq!(format!("{}", literal_ty.display(&db)), "Literal[10]");
+
+        Ok(())
+    }
+
+    fn first_public_def<'db>(db: &'db TestDb, file: File, name: &str) -> Definition<'db> {
+        let scope = module_global_scope(db, file);
+        *use_def_map(db, scope)
+            .public_definitions(symbol_table(db, scope).symbol_id_by_name(name).unwrap())
+            .first()
+            .unwrap()
+    }
+
+    #[test]
+    fn dependency_public_symbol_type_change() -> anyhow::Result<()> {
+        let mut db = setup_db();
+
+        db.write_files([
+            ("/src/a.py", "from foo import x"),
+            ("/src/foo.py", "x = 10\ndef foo(): ..."),
+        ])?;
+
+        let a = system_path_to_file(&db, "/src/a.py").unwrap();
+        let x_ty = module_global_symbol_ty_by_name(&db, a, "x");
+
+        assert_eq!(x_ty.display(&db).to_string(), "Literal[10]");
+
+        // Change `x` to a different value
+        db.write_file("/src/foo.py", "x = 20\ndef foo(): ...")?;
+
+        let a = system_path_to_file(&db, "/src/a.py").unwrap();
+
+        db.clear_salsa_events();
+        let x_ty_2 = module_global_symbol_ty_by_name(&db, a, "x");
+
+        assert_eq!(x_ty_2.display(&db).to_string(), "Literal[20]");
+
+        let events = db.take_salsa_events();
+
+        assert_function_query_was_run::<infer_definition_types, _, _>(
+            &db,
+            |ty| &ty.function,
+            &first_public_def(&db, a, "x"),
+            &events,
+        );
```

Comment:
I would remove the assertion here. The fact that `x_ty_2` resolves to a different type guarantees that the query did run.